File: app/src/main/java/com/togo/home/ui/activity/MainActivity.java
package com.togo.home.ui.activity;

import android.os.Bundle;
import android.support.v4.app.FragmentActivity;
import android.text.TextUtils;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.squareup.picasso.Picasso;
import com.togo.home.R;
import com.togo.home.data.model.SummaryWrapper;
import com.togo.home.data.remote.ServiceGenerator;
import com.togo.home.data.remote.response.PatientFirstPageModel;
import com.togo.home.support.util.CountingIdlingResourceListener;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Action;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;


public class MainActivity extends FragmentActivity {
    private final static String TAG = MainActivity.class.getSimpleName();

    @BindView(R.id.entityCell)
    protected RelativeLayout entityCell;

    @BindView(R.id.coverPhoto)
    protected ImageView coverPhoto;

    @BindView(R.id.displayName)
    protected TextView displayName;

    @BindView(R.id.emergentLine)
    protected TextView emergentLine;

    @BindView(R.id.serviceLine)
    protected TextView serviceLine;

    private Disposable disposable;
    private int ongoingId = 0;

    /// counting idle region
    private static CountingIdlingResourceListener sIdlingNotificationListener;
    public static void setIdlingNotificationListener(CountingIdlingResourceListener idlingNotificationListener) {
        sIdlingNotificationListener = idlingNotificationListener;
    }

    private static void countIdleIncrement() {
        if (sIdlingNotificationListener != null) {
            sIdlingNotificationListener.increment(); // Notify that our animation resource is busy
        }
    }

    private static void countIdleDecrement() {
        if (sIdlingNotificationListener != null) {
            sIdlingNotificationListener.decrement(); // Resource is idle again
        }
    }
    /// region end

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        tryRequest(4);
    }

    private void tryRequest(int appid) {
        if (appid == 0) {
            Log.e(TAG, "Skip invalid appid = " + appid);
            return;
        }

        if (ongoingId == appid) {
            Log.e(TAG, "Skip for ongoing request appid = " + appid);
            return;
        }

        if (null != disposable && !disposable.isDisposed()) {
            disposable.dispose();
        }

        ongoingId = appid;
        disposable = ServiceGenerator.getApiService().fetchTogoHome(appid)
                .subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .doOnTerminate(new Action() {
                    @Override
                    public void run() throws Exception {
                        ongoingId = 0;
                    }
                })
                .subscribe(new Consumer<SummaryWrapper>() {
                               @Override
                               public void accept(@NonNull SummaryWrapper apiResponse) throws Exception {
                                   handleResponse(apiResponse);
                               }
                           },
                        new Consumer<Throwable>() {
                            @Override
                            public void accept(@NonNull Throwable throwable) throws Exception {
                                handleFailure(throwable.getMessage());
                            }
                        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (null != disposable && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    private void handleResponse(SummaryWrapper apiResponse) {
        PatientFirstPageModel model = apiResponse.getData();
        if (null == model) {
            // do nothing is the right behavior?
            return;
        }

        getActionBar().setTitle(model.getDisplayName());

        if (!TextUtils.isEmpty(model.getCoverPhoto())) {
            Picasso.with(this).load(model.getCoverPhoto()).into(coverPhoto);
        }

        displayName.setText(model.getDisplayName());
        emergentLine.setText(model.getEmergentLine());
        serviceLine.setText(model.getServiceLine());

        Log.e(TAG, "Hospital name : " + model.getAboutHospitalName());
        entityCell.setVisibility(View.VISIBLE);
    }

    private void handleFailure(String message) {
        Log.e(TAG, "Error : " + message);
        Toast.makeText(MainActivity.this, message, Toast.LENGTH_LONG).show();
        entityCell.setVisibility(View.GONE);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/togo/home/ui/activity/MainActivity.kt
package com.togo.home.ui.activity

import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.support.v7.widget.LinearLayoutManager
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import com.togo.home.R
import com.togo.home.data.model.SummaryWrapper
import com.togo.home.data.retrofit.ServiceGenerator
import com.togo.home.data.retrofit.response.PatientFirstPageModel
import com.togo.home.data.source.PagesRepository
import com.togo.home.data.source.local.PagesLocalDataSource
import com.togo.home.data.source.remote.PagesRemoteDataSource
import com.togo.home.ui.util.AppFinder
import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.annotations.NonNull
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.functions.Function
import io.reactivex.functions.Predicate
import io.reactivex.schedulers.Schedulers

import kotlinx.android.synthetic.main.content_layout.*

class MainActivity : FragmentActivity() {

    private var hospitalAdapter: HospitalAdapter? = null
    private var layoutManager: LinearLayoutManager? = null

    /// AppFinder helper class region
    /// Singleton instance, subscribe with two objects: 1. a consumer, with the name seekingRefresher, to
    /// listen for app info refreshing while seeking the maximize ID. 2. an action, with the name
    /// seekingFinal, to listen for the notification of seeking complete.
    private val appFinder = AppFinder.getInstance()

    @NonNull
    private var packagesRepository: PagesRepository? = null

    @NonNull
    private var compositeDisposable: CompositeDisposable? = null

    private val seekingRefresher = Consumer<SummaryWrapper> { apiResponse ->
        // do nothing
        handleResponse(apiResponse)
    }

    private val seekingFinal = Action { rangeFetch(AppFinder.getInstance().min(), AppFinder.getInstance().max()) }
    /// AppFinder region end

    /// Fetch app first page region
    /// loop a range of app id, filtered by seekCacheChecker that skip those ids that has
    /// been refreshed before, then flatMap to request of app first page info, in the end
    /// consume by final appFirstPageConsumer.
    /// see to method 'rangeFetch'
    private val seekCacheChecker = Predicate<Int> { id -> AppFinder.getInstance().skip(id!!) }

    private val appId2FirstPageMapper = Function<Int, ObservableSource<SummaryWrapper>> { appId -> ServiceGenerator.getInstance().serviceInstance.fetchTogoHome(appId) }

    private val appFirstPageConsumer = Consumer<SummaryWrapper> { apiResponse -> handleResponse(apiResponse) }
    /// Fetch app first page region end

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        compositeDisposable = CompositeDisposable()
        packagesRepository = PagesRepository.getInstance(PagesRemoteDataSource.getInstance(),
                PagesLocalDataSource.getInstance(applicationContext))

        //        layoutManager = new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL);
        layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)

        rv.layoutManager = layoutManager
        hospitalAdapter = HospitalAdapter(this)

        rv.adapter = hospitalAdapter

        packagesRepository!!.refreshFirstPageModels()
    }

    override fun onResume() {
        super.onResume()
        fetchSavedModel()
    }

    override fun onDestroy() {
        super.onDestroy()

        compositeDisposable!!.clear()
        hospitalAdapter!!.setOnItemClickListener(null)
    }

    private fun rangeFetch(min: Int, max: Int) {
        Log.e(TAG, "rangeFetch fetching app: $min - $max")
        val disposable = Observable.range(min, max)
                .filter(seekCacheChecker)
                .flatMap(appId2FirstPageMapper)
                .subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .subscribe(appFirstPageConsumer)

        compositeDisposable!!.add(disposable)
    }


    private fun handleResponse(apiResponse: SummaryWrapper) {
        val model = apiResponse.data ?: // do nothing is the right behavior?
                return

        hospitalAdapter!!.add(model)
        updateTitle()
        saveModel(model)
        rv.scrollToPosition(hospitalAdapter!!.itemCount - 1)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        if (id == R.id.action_settings) {
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    private fun fetchSavedModel() {
        val disposable = packagesRepository!!.firstPageModels
                .subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                .subscribe { packages ->
                    if (packages.isEmpty()) {
                        appFinder.setScopeHelper(ScopeHelperImpl(applicationContext))
                        appFinder.subscribe(seekingRefresher, seekingFinal)
                    } else {
                        hospitalAdapter!!.addAll(packages)
                        updateTitle()
                    }
                }
        compositeDisposable!!.add(disposable)
    }

    private fun updateTitle() {
        title = getString(R.string.display_title, "" + hospitalAdapter!!.itemCount)
    }

    private fun saveModel(model: PatientFirstPageModel) {
        // save model to local db.
        packagesRepository!!.saveFirstPageModel(model)
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
